lib/POE/Component/Server/Twirc.pm

```diff
@@ -1150,28 +1150,76 @@ event cmd_whois => sub {
     }
 };
 
-=item notify I<on|off> I<id ...>
+=item notify I<on|off> I<screen_name ...>
 
-Turns device notifications on or off for the list of Twitter IDs.
+Turns mobile device notifications on or off for the list of I<screen_name>s.
 
 =cut
 
 event cmd_notify => sub {
-    my ($self, $channel, $argstr) = @_[OBJECT, ARG0, ARG1];
+    my $self = $_[OBJECT];
+    $self->_update_fship('device', @_[ARG0, ARG1]);
+};
+
+=item retweets I<on|off> I<screen_name ...>
+
+Turns retweet display on your timeline on or off for the list of
+I<screen_name>s.
+
+=cut
+
+event cmd_retweets => sub {
+    my $self = $_[OBJECT];
+    $self->_update_fship('retweets', @_[ARG0, ARG1]);
+};
+
+# Call update_friendships
+# All settings updated at once so existing must be preserved
+sub _update_fship {
+    my ($self, $command, $channel, $argstr) = @_;
 
     my @nicks = split /\s+/, $argstr;
     my $onoff = shift @nicks;
 
-    unless ( $onoff && $onoff =~ /^on|off$/ ) {
-        $self->bot_says($channel, "Usage: notify on|off nick[ nick [...]]");
+    unless ( $onoff && $onoff =~ /^on$|^off$/ ) {
+        $self->bot_says($channel, "Usage: $command on|off nick[ nick [...]]");
         return;
     }
 
-    my $method = $onoff eq 'on' ? 'enable_notifications' : 'disable_notifications';
+    my $setting = $onoff eq 'on' ? 1 : 0;
     for my $nick ( @nicks ) {
-        $self->twitter($method => { id => $nick });
+        # Fetch existing values
+        my %prev_vals = $self->_fetch_prev_setts($nick);
+
+        # Skip unnecessary updates
+        if ($prev_vals{$command} == $setting) {
+            $self->bot_says($channel, "No need to update $nick");
+            next;
+        }
+
+        # Update
+        $self->twitter(update_friendship =>
+                       { screen_name => $nick,
+                         # previous values as default
+                         device   => "$prev_vals{device}",
+                         retweets => "$prev_vals{retweets}",
+                         # override with new value
+                         $command => $setting
+                       });
     }
-};
+}
+
+# Fetch previous notify / retweet settings
+sub _fetch_prev_setts {
+    my ($self, $nick) = @_;
+    # Dig through twitter info
+    my $twit_data = $self->twitter(show_friendship =>
+                                  { target_screen_name => $nick })
+                        ->{relationship}{source};
+    # Pull out existing settings
+    return device   => $twit_data->{notifications_enabled},
```

Comment:
Overall unless performance is very poor (big structures) I've always seen
"pick one and be consistent" and my personal choice is always "flat or
simple" over "nested or references", I'm a bit pythonic that way :p

That said, if you'd like return by reference or if that's what's already
going on in the code base I'll change it. I did look around in the code to
try and establish a convention but didn't come up with anything

Peter
On 30 Jan 2015 16:19, "Marc Mims" <notifications@github.com> wrote:

> In lib/POE/Component/Server/Twirc.pm
> <https://github.com/semifor/twirc/pull/11#discussion_r23852175>:
>
> >      }
> > -};
> > +}
> > +
> > +# Fetch previous notify / retweet settings
> > +sub _fetch_prev_setts {
> > +    my ($self, $nick) = @_;
> > +    # Dig through twitter info
> > +    my $twit_data = $self->twitter(show_friendship =>
> > +                                  { target_screen_name => $nick })
> > +                        ->{relationship}{source};
> > +    # Pull out existing settings
> > +    return device   => $twit_data->{notifications_enabled},
>
> Returning a hashref, rather than a list, here, is better practice, I think.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/semifor/twirc/pull/11/files#r23852175>.
>